post comment

MFStevenson opened the pull request:
post works and re-renders after submit
not implemented optimistically currently as unsure how to give unque id as this happens during post

MFStevenson committed:
post comment
post works and re-renders after submit
not implemented optimistically currently as unsure how to give unque id as this happens during post

popatre reviewed (changes requested):
- nice posting functionality to the api
- Lovely controlled components for the forms

- [x] Please make the input a multitext area to allow for multi line inputs
- [x] We need at least some basic validation on the inputs, at least to stop blank posts
- [x] Need to disable the form when posting to stop duplicates or some loading pattern to stop someone spamming the post button on a slow connection
- [x] On a slow connection, theres no indication something is happening after posting, then the post just appears. 
- [x] The form at the bottom isnt great UI/UX - i have to scroll past all the comments to add a new one. If there were 1000s, this would be tedious.

MFStevenson committed:
actions feedback
now have comment box which is required
loading spinner now appears when comment is clicked
posted comment appears at bottom

nc-modou commented on src/pages/CommentsPage.jsx:
```diff
@@ -1,34 +1,73 @@
+      .then((res) => {
+        const { postedComment } = res.data;
+        setComments((currComments) => {
+          return [...currComments, postedComment];
```
Wrong order! Send the postedComment to the start of the array so it shows first

nc-modou commented on src/pages/CommentsPage.jsx:
```diff
@@ -1,34 +1,73 @@
-  const handleBackToArticle = () => {
-    navigate(`/articles/${article_id}`);
-  };
+  }, [handleSubmit]);
```
this is making 100s of requests every minute because the useEffect isn't supposed to take a function here - it doesn't even need to, because you're handling showing the new comment in handleSubmit.

nc-modou reviewed (commented):
Good response to most of Jonathan's feedback, but I noticed a few things that could do with improving
<img width="454" alt="image" src="https://github.com/MFStevenson/nc-news/assets/152851057/c322c6e8-574b-45e1-bc18-385280ef4a08">
For some reason, on the comments page there is a get request being made over and over again even when I'm not doing anything - take a look into this because the only time I should need a get request is when i first load the page

That looks like it's happening because you have a function in your dependency array on your useEffect, which is a bit of a misuse as that's designed for other data types
```
 useEffect(() => {
    getArticleComments(article_id).then((res) => {
      const { comments } = res.data;
      setComments(comments);
      setIsLoading(false);
    });
  }, [**handleSubmit**]);
```
You should remove the handleSubmit from this useEffect, because you're already handling showing that new comment in the handle submit itself

Also, the posted comment goes to the bottom which is less satisfying for the user as it's not clear if the comment was posted unless you scroll down to the bottom.

Good that you have a loading spinner while a comment is being posted now

MFStevenson committed:
action feedback
no longer making 100s of requests
comment appears at top of list

popatre reviewed (approved):
Looks all good to go 👍
